### PR TITLE
[passport-gated]: Update implementation to include Passport scoring mechanism

### PR DESCRIPTION
**Pull request by santteegt**
This PR mainly aims to update the `passport-gated` validation strategy to fully integrate with Gitcoin Passport API for validating users using a Passport stamp collection and scoring threshold. This work is a [collaboration](https://dm.raidguild.org/raids/adb5854a-94a2-4424-8a1f-c5c36b08093c) between RaidGuild and the Gitcoin Passport teams.

Changes proposed in this pull request:
- Update the `passport-gated` strategy implementation to use a scorer mechanism to calculate a Passport score for evaluating the user's eligibility.
- Update documentation to include implementation details and setup requirements.
- Update the `validation.test.ts` suite to allow evaluating multiple examples per validation strategy. Changes also enable the inclusion of true negative samples.
- Remove `passport-weighted` validation strategy from the repository as it is now deprecated.

*UPDATES*:

### Demo

[Link to Video](https://youtu.be/eJQcVPEsNWc)

Proposed changes also require to update other components in the Snapshot stack:
- https://github.com/snapshot-labs/snapshot/pull/4199
- https://github.com/snapshot-labs/score-api/pull/876

### CODE REVIEW

**Comment by plor** [comment]
I'm part of the RaidGuild team working on this and just wanted to weigh in on this to get the test to pass here. This PR adds a call to the passport API in order to score the passport to check whether it meets the threshold. To do this there are two new ENV variables required. These can be created at https://scorer.gitcoin.co/ so that the snapshot team can properly manage this secret key.

Basically the ENV variables needed are:
`PASSPORT_API_KEY` and `PASSPORT_SCORER_ID`

To create the `PASSPORT_SCORER_ID` a simple scorer can be created of the type `Unique Humanity`. I'm happy to discuss this further on the Discord server if more synchronous coordination is needed here.

**Comment by santteegt** [comment]
Hey,

Test action is failing because the PR title was changed to `[passport-weighted]` but the strategy is under `passport-gated`. Let me know if I should update my PR to change the directory name or just need to adjust the PR title.

Regarding required env variables, it would be good to set them as Action secret variables on Github.

**Comment by ChaituVR** [comment]
Hi @santteegt Thanks for the suggestions, 

for 
> To create the PASSPORT_SCORER_ID a simple scorer can be created of the type Unique Humanity

Do you mean every space will use the same `Unique humanity`?

**Comment by santteegt** [comment]
> Hi @santteegt Thanks for the suggestions,
> 
> for
> 
> > To create the PASSPORT_SCORER_ID a simple scorer can be created of the type Unique Humanity
> 
> Do you mean every space will use the same `Unique humanity`?

Yeah, the Gitcoin team requested to use the `Unique Humanity` scorer by default.

**Comment by ChaituVR** [comment]
Hi @santteegt Now that `scoreThreshold` is mandatory, it may break existing space's configurations, Is it possible to make it optional? or we should plan to announce that it won't work in the future, 

CC @zzuziak

**Comment by zzuziak** [comment]
> Hi @santteegt Now that `scoreThreshold` is mandatory, it may break existing space's configurations, Is it possible to make it optional? or we should plan to announce that it won't work in the future,
> 
> CC @zzuziak

From what I remember we agreed that the threshold for Spaces already using the validation should be set to `0` 🤔

**Comment by ChaituVR** [comment]
Could we default it to zero if undefined? 🤔

**Comment by santteegt** [comment]
> Hi @santteegt Now that `scoreThreshold` is mandatory, it may break existing space's configurations, Is it possible to make it optional? or we should plan to announce that it won't work in the future,
> 
> CC @zzuziak

you're right. I'll set the parameter as optional with zero value by default

File: .github/workflows/validation.yml
Comment (santteegt) [inline .github/workflows/validation.yml:27]: @ChaituVR added this to the workflow so it can use the Passport keys you set as secret variables

Comment (ChaituVR) [reply .github/workflows/validation.yml:27]: Yea already added them, but not sure why it doesn't read 🤔

Comment (ChaituVR) [reply .github/workflows/validation.yml:27]: Works well in local

**Comment by zzuziak** [comment]
Hey @santteegt, amazing work 🔥  Could you also possibly share screenshots or a short Loom recording with the Validation in UI?

File: .github/workflows/validation.yml
Comment (santteegt) [reply .github/workflows/validation.yml:27]: tested locally using [https://github.com/nektos/act](act) and it's working. There might be an issue with how you're setting the secret env vars in the repo settings

**Comment by santteegt** [comment]
> Hey @santteegt, amazing work 🔥 Could you also possibly share screenshots or a short Loom recording with the Validation in UI?

Sure! But don't know how to load the updated strategies on a local snapshot frontend though. I was exploring the frontend repo on how to do that but found that the latest validation strategies metadata is being fetched from a [https://score.snapshot.org/api/validations](API endpoint). Could you point me to a good guide on how to setup both the frontend and strategies locally?

**Comment by santteegt** [comment]
Hey @zzuziak 

I added a link to the demo in the PR description. While recording the video I noticed that we also required to update other components in your stack. More specifically, the Scorer API and the frontend. I already submitted a couple PRs with those changes, you can find them listed at the top of this PR as well

File: src/validations/passport-gated/index.ts
Comment (ChaituVR) [inline src/validations/passport-gated/index.ts:203]: HI @santteegt This should not be called if there is no `scoreThreshold`

Comment (ChaituVR) [reply src/validations/passport-gated/index.ts:203]: This address `0x24F15402C6Bb870554489b2fd2049A85d75B982f` got ENS stamp but it still returns `false` because of the passport score 🤔

File: src/validations/passport-gated/index.ts
Comment (ChaituVR) [inline src/validations/passport-gated/index.ts:175]: If we still wait for scorer after 2 attempts, we should return an error instead of `false`

**Review by ChaituVR — CHANGES_REQUESTED** [review]
So for `validStamps` we are checking snapshot block time to check if user has stamps only before proposal creation, but for `validatePassportScore` we are not checking any time related score,  maybe we should make it optional for validStamps (default to check time)

**Comment by santteegt** [comment]
> So for `validStamps` we are checking snapshot block time to check if user has stamps only before proposal creation, but for `validatePassportScore` we are not checking any time related score, maybe we should make it optional for validStamps (default to check time)

Checking passport stamps validity is required to ensure a member has a valid issued stamp before proposal creation / proposal voting timesstamp. Passport score however could be calculated at any time by independent Scorers (with `PASSPOT_SCORER_ID`) so it's not relevant to eval `last_score_timestamp` as soon the passport stamps have been issued before a validation event

**Comment by 0xAurelius** [comment]
Once this is finalized and merged, would it be reasonable for me to open a follow-up PR to add a "at-least-n" operator option and corresponding "n" parameter?

The use-case is that the DAO I contribute to would like to require at least n=2 of the approved stamps (regardless of Score)...

**Comment by 0xZakk** [comment]
Hey @ChaituVR, wondering what next steps are required to get this finalized and merged?

**Comment by ChaituVR** [comment]
I think we are good with this. we planning to release this tomorrow (20th sept)

**Comment by ChaituVR** [comment]
FYI, just waiting for a small doubt to be answered in our telegram chat, will merge it soon after that :)

**Comment by 0xZakk** [comment]
@santteegt - while testing, @ChaituVR ran into an issue where he updated his score in the Passport UI and then still wasn't able to vote with the strategy. We need to submit the address to be scored, then fetch the score. Even though this ends up hitting the API a lot, I think the best approach might be to submit the score each time (which will often give you the score back immediately, in the response object)

**Comment by santteegt** [comment]
@0xZakk @ChaituVR thanks for the feedback. I made a quick update on the implementation to consider that scenario. The PR is ready for another review

**Comment by 0xZakk** [comment]
@santteegt thank you! That was fast! Looks good to me. The work day is finished for @ChaituVR, so he'll review it by tomorrow morning (our time)

File: src/validations/passport-gated/README.md
Comment (ChaituVR) [inline src/validations/passport-gated/README.md:72]:
```suggestion
This code was last modified on September 21, 2023.
```

File: src/validations/passport-gated/index.ts
Comment (ChaituVR) [inline src/validations/passport-gated/index.ts:109]: Threshold should include minimum value also right?? like if `scoreThreshold` is `17.62`, any validation with a score of `17.61` or lower would be considered invalid, while any validation with a score of `17.62` or higher would be considered valid. no?

here it is checking from `17.63` and more to be valid

```suggestion
  return Number(scoreData.score) >= minimumThreshold;
```

File: src/validations/passport-gated/index.ts
Comment (ChaituVR) [inline src/validations/passport-gated/index.ts:140]: This API will return only DONE and PROCESSING? 🤔

**Review by ChaituVR — CHANGES_REQUESTED** [review]
Everything else looks good :) 👍 Thanks for the quick fixes

File: src/validations/passport-gated/index.ts
Comment (santteegt) [reply src/validations/passport-gated/index.ts:140]: right. As it always hits the `/submit-passport` endpoint first, the `GET_PASSPORT_SCORE_URI` will be called only if the former returns `PROCESSING` status

**Review by ChaituVR — APPROVED** [review]
Thank you ❤️ 


@zzuziak  will take care of documentations and announcements very soon :)

**Pull request merged**
